104.md

```diff
@@ -0,0 +1,205 @@
+    "tags": [
+        ["p", <pubkey of the recipient>],
+        ["prekey", <pubkey of the recipient prekey used for DH calculation>],
+        ["ephemeral", <pubkey of the ephemeral key used for DH calculation>]
+    ],
+    "sig": "" // No signature because these are only sent inside gift wraps.
+}
+```
+
+### Double Ratchet Direct Message (DRDM) Event
+
+The Double Ratchet Direct Message (DRDM) event is the main event kind that carries the conversation content. This event carries enough information for recievers to properly double ratchet their chains and decrypt the message content with the right symmetrical keys.
+
+```json
+{
+    "kind": 444,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey of the sender>,
+    "content": <ciphertext of the message symmetrically encrypted with chain/message key>,
+    "tags": [
+        ["p", <pubkey of the recipient>],
+        ["dh_sending", <current DH sending pubkey>],
+        ["current_index", <number of current message in the current message chain>],
+        ["previous_length", <length of previous message chain>]
+    ],
+    "sig": "" // No signature because these are only sent inside gift wraps.
+}
+```
+
+#### Required Tags
+
+- `"p"`: References the recipient
+- `"dh_sending"`: The ephemeral pubkey used in the current Diffie-Hellman ratchet.
+- `"current_index"`: The index of the current message in the current message chain. For handling out of order messages.
+- `"previous_length"`: The length of the previous message chain. For handling out of order messages.
+
+Clients can use any other standard tags they like, for example, `"a"` or `"e"` tags to reference previous messages.
+
+### Prekey Event
+
+The Prekey event is a simple replaceable event where a user publishes their signed prekey that is required to initiate the Diffie-Hellman key exchange to create a conversation request event. Including a signature on this event proves that the holder of the user's main nostr key (which signs the `kind: 10443` event) also has control of the private key associated with the prekey.
+
+In most cases, clients that implement this NIP will manage the creation and rotation of this key on a regular basis. It's recommended that clients do so interactively with user consent in order to avoid overwriting prekeys created by other clients. It is also recommended that clients rotate this key on a regular basis.
+
+```json
+{
+    "kind": 10443,
+    "created_at": <unix timestamp in seconds>,
+    "pubkey": <pubkey>,
+    "content": <hex encoded pubkey of the prekey>,
+    "tags": [
+        ["prekey_sig", <signature generated from hex encoded pubkey of the prekey>]
+    ],
+    "sig": <signed normally with the user's main nostr pubkey>
+}
+```
+
+The `"prekey_sig"` tag value is a Schnorr signature of the sha256 hashed value of the prekey's pubkey, signed with the prekey's private key.
```

Comment:
Clients MUST verify the `prekey_sig` before accepting this event. 

This allows anyone out there to copy the `prekey_sig` string and resign it under their own main nostr keys. Shouldn't the signature be a hash of `(nostr.pubkey || prekey.pubkey)` to avoid the shenanigans of just reusing somebody else's key?

Reply:
I don't know. As it is today, one can just copy any other user's prekey tag and it will be valid. They won't have the private key so maybe they can't do anything with it, but there might be a social attack somewhere. The event doesn't prove the prekey is linked with the main key.

Reply:
True. that's a good point. I was thinking validation of the event + validation of the prekey signature would be enough but you're right, they aren't technically linked and it's easy to link them. 

I don't think it's strictly necessary given how we do the multiple pairs of DH calculations in the initial key exchange (which does require private key signatures from both the prekey and the identity key) to derive the initial root key but it certainly wouldn't hurt.